packages/flame_studio/lib/src/widgets/panels/hierarchy_view.dart

```diff
@@ -0,0 +1,221 @@
+              color: Color(0xffffffff),
+            ),
+          ),
+        ),
+      ),
+    );
+    final punctuation = Transform(
+      transform: Matrix4.translationValues(-3.0, 0, 0),
+      child: Text(
+        hasChildren ? (isExpanded ? ' {' : ' {...}') : ',',
+        style: const TextStyle(color: Color(0x66f5d49a)),
+      ),
+    );
+
+    out.add(
+      Row(
+        children: [
+          expanderIcon,
+          indent,
+          componentName,
+          punctuation,
+        ],
+      ),
+    );
+    if (isExpanded && hasChildren) {
+      for (final childNode in node.children!) {
+        _buildList(childNode, depth + 1, out);
+      }
+      out.add(
+        Row(
+          children: [
+            _ExpanderIcon(hasChildren: false, isLast: depth == 0),
+            SizedBox(width: 15.0 * depth + 5.0),
+            const Text(
+              '}',
+              style: TextStyle(color: Color(0x66f5d49a)),
+            ),
+          ],
+        ),
+      );
+    }
+  }
+
+  /// Used by the expander icons in the gutter: the component expands/collapses
+  /// without affecting the selection state.
+  void _toggle(Component component) {
+    setState(() {
+      if (expandedComponents.contains(component)) {
+        expandedComponents.remove(component);
+      } else {
+        expandedComponents.add(component);
+      }
+    });
+  }
+
+  void _toggleOrSelect(Component component) {
+    final isExpanded = expandedComponents.contains(component);
+    final isSelected = selectedComponent == component;
+    setState(() {
+      if (isExpanded && isSelected) {
+        expandedComponents.remove(component);
+      } else {
+        expandedComponents.add(component);
+      }
+      selectedComponent = component;
+    });
+  }
+}
+
+class _ExpanderIcon extends StatelessWidget {
+  const _ExpanderIcon({
+    required this.hasChildren,
+    this.isExpanded = false,
+    this.isFirst = false,
+    this.isLast = false,
+  });
+  final bool hasChildren;
+  final bool isExpanded;
+  final bool isFirst;
+  final bool isLast;
+
+  @override
+  Widget build(BuildContext context) {
+    return CustomPaint(
+      size: const Size(15, 20),
+      painter: _ExpanderIconPainter(this),
+    );
+  }
+}
+
+class _ExpanderIconPainter extends CustomPainter {
+  _ExpanderIconPainter(this.icon);
+
+  final _ExpanderIcon icon;
+  static final _paint = Paint()
+    ..color = const Color(0xFFFFFFFF)
+    ..style = PaintingStyle.stroke;
+
+  @override
+  bool shouldRepaint(_ExpanderIconPainter old) =>
+      (icon.hasChildren != old.icon.hasChildren) ||
+      (icon.isExpanded != old.icon.isExpanded);
+
+  @override
+  void paint(Canvas canvas, Size size) {
+    const iconSize = 10.0;
+    const halfSize = iconSize / 2;
+    final xExtent = size.width / 2;
+    final yExtent = size.height / 2;
+    canvas.save();
+    canvas.translate(xExtent, yExtent);
+    if (icon.hasChildren) {
+      if (!icon.isFirst) {
+        canvas.drawLine(
```

Comment:
We could use `..` cascading here on the consecutive calls.

Reply:
Most of these calls have if conditions, so we wouldn't be able to cascade them.